File: longest_palindromic_substring.cpp
#include "header.h"

string expand(string s, int i, int j){
    while(i>=0 && j<=s.size()-1 && s[i] == s[j]){
        i--;
        j++;
    }
    return s.substr(i+1,j-1-i);
}

string longestPalindrome(string s) {
    if(s.size() == 0) return "";
    string res = s.substr(0,1);
    for(int i=0; i<s.size(); i++){
        string x = expand(s,i,i);
        if(x.size()>res.size())
            res = x;
    }
    for(int i=0; i<s.size()-1; i++){
        string x = expand(s,i,i+1);
        if(x.size()>res.size())
            res = x;
    }
    return res;
}

int main(){
    string s0 = longestPalindrome("abacdfgdcaba");
    printf("%s\n", s0.c_str());
    return 0;
}

File: longest_palindromic_substring.py
class Solution:
    def expand(self,s,i,j):
        l,r = i,j
        while l>=0 and r<len(s):
            if s[l] == s[r]:
                l-=1
                r+=1
            else:
                break
        return s[l+1:r]
    # @return a string
    def longestPalindrome(self, s):
        if len(s)<2:
            return s
        s_len = len(s)
        res = s[:1]
        for i in range(0,s_len,1):
            temp_str = self.expand(s,i,i)
            if len(temp_str)>len(res):
                res = temp_str
        for i in range(0,s_len-1,1):
            temp_str = self.expand(s,i,i+1)
            if len(temp_str)>len(res):
                res = temp_str
        return res

print Solution().longestPalindrome("abacdfgdcaba")
